_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,229 @@
+    hiBits : int [@bs.as "hi]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+It will generate JS output as below:
+```js
+var value = {lo : 32, hi : 33}
+function rand (param){
+    return param.lo + param.hi
+}
+```
+
+If users want to make it even  shorter, they can choose to compile record as an array as below:
+
+```reasonml
+type int64 = {
+    loBits : int [@bs.as "0"],
+    hiBits : int [@bs.as "1"]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+Now output JS as below:
+
+```js
+var value = [32,33]
+function rand(param){
+    return param[0] + param[1]
+}
+```
+
+The label renaming techniques can be applied systematically using a syntactic macro, in the future we may provide an advanced mode to apply it automatically. Another nice property is that only the type definition needs to be adapted, other parts of code is untouched.
+
+### IC friendliness
+
+Record is always in a perfect IC position, the compiler can ensure all generated records are of the same shape
+
+## Variant (internal)
+This encoding for variants may be  subject to change in the future, but it is so simple that it makes sense for users to have a basic understanding.
+
+Take such type definition for example:
+
+```reasonml
+type t =
+  | Black(t, int, t)
+  | Red(t, int, t)
+  | Empty;
+let empty = Empty  ;
+let v0 = Black (empty, 3, empty);
+let v1 = Red (empty, 3, empty);
+```
+The generated JS code would be:
+
+```js
+var empty =/*Empty*/ 0;
+var v0  = {TAG : 0/*Black*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+var v1 = {TAG : 1/*Red*/, _0 : /*Empty*/ 0 , _1 : 3 , _2 : /*Empty */ 0};
+```
+
+As you can see, variants are divided into two categories, the variant which does not have payload is compiled into a number starting from 0, while variants which has the payload is compiled into an object which has the first slot named `TAG` and the following slots named as `_0`, `_1` ..
+
+### variant with inline records
+
+Users can give names to the payload, and the compiler respect it, however, we don't support user level renaming, i.e, using `bs.as`, at this time.
+
+```reasonml
+type t =
+  | Black ({l:t, value: int, r: t})
+  | Red({l:t, value: int, r: t})
+  | Empty;
+let empty = Empty  ;
+let v0 = Black ({l:empty, value: 3, r:empty});
+let v1 = Red ({l:empty, value:3, r: empty});
+```
+The generated JS code would be:
+```js
+var empty =/*Empty*/ 0;
+var v0  = {TAG : 0/*Black*/, l: /*Empty*/ 0 , value : 3 , r: : /*Empty */ 0};
+var v1 = {TAG : 1/*Red*/, l : /*Empty*/ 0 , value : 3 , r : /*Empty */ 0};
+``` 
+
+
+### Special case when the number of variants which has payload is only 1.
+
+Take the types below for example:
+
+```reasonml
+type list = 
+    | Nil
+    | Cons (int * list);
+```
+
+Since the number of variants which has payload is only 1, the compiler does not need add `TAG` when we destruct the data for pattern matching, so the code below:
```

Comment:
Since only one variant has a payload, the compiler does not need to add `TAG` when we destructure the data for pattern matching. Thus, the following code: